fix: FalkorDB profile failing to start in docker-compose

Naseem77 opened the pull request:
- Remove --system flag to install packages to venv where app runs
- Add --no-sync to CMD to preserve build-time installations
- Make Neo4j settings optional, add FalkorDB settings
- Add conditional driver initialization based on db_backend

## Summary

1. Dockerfile (2 changes)
Change 1: Removed --system flag

before: uv pip install --system --upgrade "graphiti-core[falkordb]"
after:  uv pip install --upgrade "graphiti-core[falkordb]"

* --system installs to system Python (/usr/local/lib/python3.12/)
* Without --system, installs to venv (/app/.venv/lib/python3.12/)
* App runs with venv in PATH, so packages must be in venv 

Change 2: Added --no-sync to CMD

before: CMD ["uv", "run", "uvicorn", ...]
after: CMD ["uv", "run", "--no-sync", "uvicorn", ...]

* uv run normally re-syncs venv from lockfile on every container start
* --no-sync skips re-sync, preserving build-time installations
* Without this, FalkorDB packages installed at build would be overwritten at runtime

  2. server/graph_service/config.py
Made Neo4j fields optional, added FalkorDB fields

* Allows settings to load without Neo4j credentials when using FalkorDB
* db_backend determines which database driver to use
* Backward compatible: Neo4j remains default

 3. server/graph_service/zep_graphiti.py
Updated __init__ to accept optional parameters

* Accepts optional Neo4j parameters (matching config.py)
* **kwargs captures graph_driver parameter for FalkorDB 

Added conditional driver initialization
if settings.db_backend == 'falkordb':
    driver = FalkorDriver(host=..., port=..., database=...)
    client = ZepGraphiti(graph_driver=driver)
else:
    client = ZepGraphiti(uri=..., user=..., password=...) 
 
* Checks db_backend to determine which database to use
* Creates FalkorDriver for FalkorDB, uses Neo4j parameters otherwise
* FalkorDriver imported locally only when needed


## Type of Change
- [x] Bug fix
- [ ] New feature
- [ ] Performance improvement
- [ ] Documentation/Tests

## Objective
**For new features and performance improvements:** Clearly describe the objective and rationale for this change.

## Testing
- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] All existing tests pass

## Breaking Changes
- [ ] This PR contains breaking changes

If this is a breaking change, describe:
- What functionality is affected
- Migration path for existing users

## Checklist
- [ ] Code follows project style guidelines (`make lint` passes)
- [ ] Self-review completed
- [ ] Documentation updated where necessary
- [ ] No secrets or sensitive information committed

## Related Issues
Closes #749

Naseem77 committed:
fix: FalkorDB profile failing to start in docker-compose

- Remove --system flag to install packages to venv where app runs
- Add --no-sync to CMD to preserve build-time installations
- Make Neo4j settings optional, add FalkorDB settings
- Add conditional driver initialization based on db_backend

Naseem77 committed:
fix: add type ignores for FalkorDB driver initialization

Naseem77 committed:
fix: add type ignores for FalkorDriver calls and parent init

Naseem77 committed:
fix: suppress pyright errors for FalkorDB initialization

Naseem77 committed:
fix: add type ignore to each FalkorDriver parameter line

prasmussen15 reviewed (commented):
## Code Review

Thanks for this PR! The Dockerfile fixes and FalkorDB support additions look good. I have a few suggestions to improve the code:

### 1. Code Duplication - Extract Helper Function

The FalkorDB driver initialization is duplicated in both `get_graphiti()` and `initialize_graphiti()`. Consider extracting into a helper:

```python
def _create_graphiti_client(settings: ZepEnvDep) -> ZepGraphiti:
    """Create a ZepGraphiti client based on the configured database backend."""
    if settings.db_backend == 'falkordb':
        from graphiti_core.driver.falkordb_driver import FalkorDriver

        driver = FalkorDriver(
            host=settings.falkordb_host or 'localhost',
            port=settings.falkordb_port or 6379,
            database=settings.falkordb_database or 'default_db',
        )
        return ZepGraphiti(graph_driver=driver)
    else:
        # Validate Neo4j settings are present
        if not all([settings.neo4j_uri, settings.neo4j_user, settings.neo4j_password]):
            raise ValueError(
                "Neo4j configuration (neo4j_uri, neo4j_user, neo4j_password) is required "
                "when db_backend is 'neo4j'"
            )
        return ZepGraphiti(
            uri=settings.neo4j_uri,
            user=settings.neo4j_user,
            password=settings.neo4j_password,
        )
```

### 2. Resource Leak - Missing `client.close()` in `initialize_graphiti()`

`get_graphiti()` properly closes the client in a `finally` block, but `initialize_graphiti()` doesn't close the client after `build_indices_and_constraints()`. This could leak database connections:

```python
async def initialize_graphiti(settings: ZepEnvDep):
    client = _create_graphiti_client(settings)
    try:
        await client.build_indices_and_constraints()
    finally:
        await client.close()
```

### 3. Add Validation for Neo4j Backend

When `db_backend` is `'neo4j'` (default), the code should validate that required Neo4j settings are present rather than passing `None` values silently. (This is included in the helper function above.)

---

With these changes, `get_graphiti()` and `initialize_graphiti()` become much simpler:

```python
async def get_graphiti(settings: ZepEnvDep):
    client = _create_graphiti_client(settings)
    if settings.openai_base_url is not None:
        client.llm_client.config.base_url = settings.openai_base_url
    if settings.openai_api_key is not None:
        client.llm_client.config.api_key = settings.openai_api_key
    if settings.model_name is not None:
        client.llm_client.model = settings.model_name

    try:
        yield client
    finally:
        await client.close()


async def initialize_graphiti(settings: ZepEnvDep):
    client = _create_graphiti_client(settings)
    try:
        await client.build_indices_and_constraints()
    finally:
        await client.close()
```

The Dockerfile changes and config.py changes look good! 👍

prasmussen15 reviewed (changes requested):
Please address the comments

Naseem77 committed:
Refactor graphiti client initialization per code review

- Extract _create_graphiti_client() helper to eliminate duplication
- Fix resource leak: add client.close() in initialize_graphiti()
- Add validation for Neo4j configuration when db_backend='neo4j'"

Naseem77 committed:
fix pyright warning

Naseem77 commented:
@prasmussen15 updated the comments.

prasmussen15 reviewed (approved):
App-layer changes verified locally — image builds with the venv install + `--no-sync`, server boots, FalkorDriver connects, and /healthcheck returns healthy. Merging; the remaining docker-compose port mismatch (8001→8000) will be addressed in a fast-follow PR.
